### PR TITLE
feat(node | das | libs/header/sync): add total uptime node metrics + totalSampled das metrics + totalSynced sync metrics

### PR DESCRIPTION
**Pull request by derrandz**
## Overview

This PR introduces node uptime metrics + das total sampled headers metrics to support calculating the uptime index proposed by mustafa on the monitoring side.

This PR introduces a new module named `Telemetry` to support node related telemetry. This module can also host all general telemetry and observability that does not interest specific modules.

## Changes

- [x] Introduced uptime metrics for node under `nodebuilder/node/uptime.go`
- [x] Introduced persistent uptime metrics using datastore to persist node start time
- [x] Testing for uptime metrics persistence using the store
- [x] Unit testing for uptime metrics
- [x] Integration testing for uptime metrics
- [ ] e2e testing for uptime metrics

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [ ] Linked issues closed with keywords

## Blocked By

PR: #1537

### CODE REVIEW

File: nodebuilder/node/uptime.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// This file defines UptimeMetrics relative to the nodebuilder package.
```
Comment (renaynay) [inline nodebuilder/node/uptime.go:1]:
```suggestion
```

File: nodebuilder/node/uptime.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
+)
+
+// UptimeMetrics is a struct that records
+// 1. node start time: the timestamp when the node was started
+// 2. node up time: total time the node has been running
+//
+// the node start time is recorded imperatively when RecordNodeStartTime is called
+// whereas the node up time is recorded periodically upon callback recalling (re-mettering from optl)
```
Comment (renaynay) [inline nodebuilder/node/uptime.go:19]:
```suggestion
// whereas the node uptime is recorded periodically upon callback recalling (re-metering from optl)
```

File: nodebuilder/node/uptime.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+	"go.opentelemetry.io/otel/metric/instrument/syncfloat64"
+)
+
+// UptimeMetrics is a struct that records
+// 1. node start time: the timestamp when the node was started
+// 2. node up time: total time the node has been running
```
Comment (renaynay) [inline nodebuilder/node/uptime.go:16]:
```suggestion
// 2. node uptime: total time the node has been running
```

File: nodebuilder/node/uptime.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	// lastNodeUptimeTS is the last timestamp when the node uptime was recorded.
+	lastNodeUptimeTS float64
+}
+
+var meter = global.MeterProvider().Meter("node")
+
+// NewUptimeMetrics creates a new UptimeMetrics
+// and registers a callback to re-meter the totalNodeUptime metric.
+func NewUptimeMetrics() (*UptimeMetrics, error) {
+	nodeStartTS, err := meter.
```
Comment (mojtaba-esk) [inline nodebuilder/node/uptime.go:36]: Does this store the node start value in some sort of db and retrieve it if available? because if it doesn't, once the node is restarted, we get a new start time for the node which yields in a wrong uptime calculation.

**Review by mojtaba-esk — CHANGES_REQUESTED** [review]

File: das/metrics.go [hunk 5]
```diff
@@ -76,11 +78,22 @@ func (d *DASer) InitMetrics() error {
 		return err
 	}
 
+	totalSampled, err := meter.
+		SyncInt64().
+		UpDownCounter(
```
Comment (walldiss) [inline das/metrics.go:83]: Consider using gauge. Counter will start from previous value upon restart, that will break observations

File: das/metrics.go [hunk 6]
```diff
@@ -110,29 +123,43 @@ func (d *DASer) InitMetrics() error {
...
 
+// observeSample records the time it took to sample a header +
+// the amount of sampled contiguous headers
 func (m *metrics) observeSample(ctx context.Context, h *header.ExtendedHeader, sampleTime time.Duration, err error) {
 	if m == nil {
 		return
 	}
 	m.sampleTime.Record(ctx, sampleTime.Seconds(),
 		attribute.Bool("failed", err != nil),
-		attribute.Int("header_width", len(h.DAH.RowsRoots)))
```
Comment (walldiss) [inline das/metrics.go:119]: Why did you remove attribute?

File: das/metrics.go [hunk 7]
```diff
@@ -110,29 +123,43 @@ func (d *DASer) InitMetrics() error {
...
 	}
 	m.newHead.Add(ctx, 1)
 }
+
+// recordTotalSampled records the total amount of sampled headers.
+func (m *metrics) recordTotalSampled(ctx context.Context, n int64) {
+	if m == nil {
+		return
+	}
+	m.totalSampled.Add(ctx, n)
```
Comment (walldiss) [inline das/metrics.go:164]: Oh man, the counter could not be used like that, you need gauge. I won't be long until you observe sampled percentage to be > 100%

File: nodebuilder/node/uptime.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+// 2. node up time: total time the node has been running
+//
+// the node start time is recorded imperatively when RecordNodeStartTime is called
+// whereas the node up time is recorded periodically upon callback recalling (re-mettering from optl)
+type UptimeMetrics struct {
+	// nodeStartTS is the timestamp when the node was started.
+	nodeStartTS syncfloat64.UpDownCounter
+
+	// totalNodeUptime is the total time the node has been running.
+	totalNodeUptime asyncfloat64.Gauge
```
Comment (walldiss) [inline nodebuilder/node/uptime.go:25]: This could benefit from using a single counter. Add 1 every second node is running. It will give you run indicater + sum could be used as uptime in seconds later. Gauge is bad here, the value could be calculated by aggregate. 

Start time on the other hand should be a Gauge. Do you even need start time? You can do uptime calculation with it.

File: das/coordinator.go [hunk 9]
```diff
@@ -55,6 +55,12 @@ func newSamplingCoordinator(
 
 func (sc *samplingCoordinator) run(ctx context.Context, cp checkpoint) {
 	sc.state.resumeFromCheckpoint(cp)
+
+	// the amount of sampled headers from the checkpoint
+	// is the point from which we should start sampling - 1
+	totalSampledFromCheckpoint := int64(cp.SampleFrom - 1)
```
[the remark this answers is not in the document]
Comment (derrandz) [reply das/coordinator.go:61]: Got it

File: das/metrics.go (hunk 6 above)
Comment (derrandz) [reply das/metrics.go:119]: a relic

File: nodebuilder/node/uptime.go (hunk 8 above)
Comment (derrandz) [reply nodebuilder/node/uptime.go:25]: Turning this into a counter adding 1 every other second is a great idea! Thank you
On the other hand, I think we need to record the node start time timestamp to achieve the uptime index formula from Mustafa.

File: nodebuilder/node/uptime.go (hunk 4 above)
Comment (derrandz) [reply nodebuilder/node/uptime.go:44]: Good point, now implemented a mechanism to store it in the datastore in the [following commit](https://github.com/celestiaorg/celestia-node/pull/1638/commits/f01f9281292bc8926691f98e330177c33b8a4bda)

**Comment by derrandz** [comment]
We're successfully retrieving metrics at the moment, however we need to think about calculating the uptime index (from mustafa's proposal) somewhere either in node or in mojtaba's API, since Grafana's min function returns a min over a vector and not a minimum of two values.

Here's a screen short of the two values, green being `totalSampled/network_head`, and yellow being `(time-nodeStartTs)/total_uptime_in_s`

Screen Shot 2023-01-27 at 14 56 12

File: das/coordinator.go [hunk 10]
```diff
@@ -76,6 +81,12 @@ func (sc *samplingCoordinator) run(ctx context.Context, cp checkpoint) {
 			}
 		case res := <-sc.resultCh:
 			sc.state.handleResult(res)
+
+			// totalSampledFromWorker is the amount of successfully
+			// sampled headers from the worker
+			totalSampledFromWorker := int64(res.To-res.From) - int64(len(res.failed))
+			sc.metrics.recordTotalSampled(ctx, totalSampledFromWorker)
```
Comment (walldiss) [inline das/coordinator.go:88]: Async metrics could only be observed inside metrics Callback. Consider storing a int64 value inside metrics struct and observe it on callback.

File: das/checkpoint.go [hunk 11]
```diff
@@ -48,3 +48,17 @@ func (c checkpoint) String() string {
...
 }
+
+// WorkersTotalSampled returns the total amount of sampled headers
+func (c checkpoint) WorkersTotalSampled() int {
+	var total int
+	for _, w := range c.Workers {
+		total += int(w.To - w.From)
+	}
+	return total
+}
```
Comment (walldiss) [inline das/checkpoint.go:59]: Name of the function doesn't match the logic. You calculating amount of not sampled items, that workers will pick up upon resume. Those headers are not yet sampled.

Also if w.To == w.From there is 1 header to be sampled. So total should be like this:
```
total += int(w.To - w.From) + 1
```

File: das/checkpoint.go [hunk 12]
```diff
@@ -48,3 +48,17 @@ func (c checkpoint) String() string {
...
+
+// TotalSampled returns the total amount of sampled headers
+func (c checkpoint) TotalSampled() int {
+	return int(c.SampleFrom) - 1 + c.WorkersTotalSampled() - len(c.Failed)
+}
```
Comment (walldiss) [inline das/checkpoint.go:64]: TotalSampled and WorkersTotalSampled could be non-exported. Also consider making 1 function onstead of 2, there is no need to have them both in API surface.

File: nodebuilder/node/uptime.go [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			totalNodeUptime,
+		},
+		func(ctx context.Context) {
+			m.totalNodeUpTimeTicks++
+			totalNodeUptime.Observe(ctx, float64(m.totalNodeUpTimeTicks))
+		},
+	)
```
Comment (walldiss) [inline nodebuilder/node/uptime.go:83]: Looks like you are observing amount of callback calls here, but not the amount of seconds from node start.

**Review by renaynay — COMMENTED** [review]
just define the const in `uptime.go`

File: nodebuilder/node/uptime.go (hunk 13 above)
Comment (derrandz) [reply nodebuilder/node/uptime.go]: Correct, noticed this while visually testing this. Fixed it, thank you.

File: das/checkpoint.go (hunk 11 above)
Comment (derrandz) [reply das/checkpoint.go:59]: I'm slightly confused, what you are saying regarding the "to be sampled" ranges for workers, makes sense, however in your [previous comment](https://github.com/celestiaorg/celestia-node/pull/1638#discussion_r1087757155) you mentioned that `w.To - w.From` would be the all done from workers.

So which is correct, `c.SampleFrom - 1 - allFailed` or should add to these the workers' to and from? I'm thinking the first formula.

Comment (walldiss) [reply das/checkpoint.go:59]: The comment you mentioned doesn't say w.To - w.From would be the all done. It is items that were not finished, and will be continued after restart by resumed workers.  Thus sum of those needs to be subtracted from current allSampled value.

File: das/coordinator.go (hunk 10 above)
Comment (derrandz) [reply das/coordinator.go:88]: 👍🏻

File: das/checkpoint.go (hunk 11 above)
Comment (derrandz) [reply das/checkpoint.go:59]: Thank you for the elaboration @walldis, as you've helped point out through discussion, the correct formula is:

```
totalSampled = c.SampleFrom - TotalInProgress - TotalFailed
```

**Comment by derrandz** [comment]
@mojtaba-esk and I discussed the following [comment](https://github.com/celestiaorg/celestia-node/pull/1638#discussion_r1087714225) and concluded to not store any metrics on store at the node side, and defer this requirement to the parties that require it (_i.e: NodeLogger and Leaderboard backend_)

File: das/coordinator.go [hunk 14]
```diff
@@ -55,6 +55,11 @@ func newSamplingCoordinator(
 
 func (sc *samplingCoordinator) run(ctx context.Context, cp checkpoint) {
 	sc.state.resumeFromCheckpoint(cp)
+
+	// the amount of sampled headers from the last checkpoint
+	totalSampledFromCheckpoint := int64(cp.totalSampled())
+	sc.metrics.recordTotalSampled(ctx, totalSampledFromCheckpoint)
```
Comment (walldiss) [inline das/coordinator.go:61]: Nit: no need for extra variable

File: das/coordinator.go (hunk 10 above)
Comment (walldiss) [inline das/coordinator.go:88]: resultsCh is working with sampling ranges. There is`observeSample` func, that is called every time block is sampled.  You could get much more granular results by updating totalSampled inside `observeSample`.  Would also cleanup coordinator a bit.

File: das/metrics.go [hunk 15]
```diff
@@ -110,29 +132,46 @@ func (d *DASer) InitMetrics() error {
...
 	m.newHead.Add(ctx, 1)
 }
+
+// recordTotalSampled records the total amount of sampled headers.
+func (m *metrics) recordTotalSampled(ctx context.Context, n int64) {
+	if m == nil {
+		return
+	}
+	totalSampledInt := atomic.LoadInt64(&m.totalSampledInt)
+	atomic.StoreInt64(&m.totalSampledInt, totalSampledInt+n)
```
Comment (walldiss) [inline das/metrics.go:176]: Two subsequent atomic operation combined does not have transaction property. Please use `atomic.AddInt64(&m.totalSampledInt,1)`

File: nodebuilder/settings.go [hunk 16]
```diff
@@ -83,7 +92,7 @@ func initializeMetrics(
 			exp,
 		),
 		controller.WithExporter(exp),
-		controller.WithCollectPeriod(2*time.Second),
+		controller.WithCollectPeriod(node.OptlCollectPeriodInSeconds*time.Second),
```
Comment (walldiss) [inline nodebuilder/settings.go:95]: collect period affects load on collector and a lot of other things, it should not be set based on one logic of one particular metric. Please remove

File: nodebuilder/node/uptime.go [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+		totalNodeUpTimeTicks: 1,
+	}
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			totalNodeRunTime,
+		},
+		func(ctx context.Context) {
+			m.totalNodeUpTimeTicks += OptlCollectPeriodInSeconds
+			totalNodeRunTime.Observe(ctx, float64(m.totalNodeUpTimeTicks))
```
Comment (walldiss) [inline nodebuilder/node/uptime.go:76]: consider storing last observation time and collecting seconds by using something like `time.Since(lastObservationTime).Seconds()`

File: nodebuilder/settings.go (hunk 16 above)
Comment (derrandz) [reply nodebuilder/settings.go:95]: It's the other way around, the metric is depending on this value. So we decided to put this hard-coded value into a constant so that it's accessible elsewhere.

File: nodebuilder/node/uptime.go (hunk 17 above)
Comment (walldiss) [reply nodebuilder/node/uptime.go:76]: You need to increment the counter by amount of seconds passed since last observation (time.Since().Seconds()). You don't need to have any constants for that, nor modifying internals of collector.

File: nodebuilder/settings.go (hunk 16 above)
Comment (walldiss) [reply nodebuilder/settings.go:95]: Lets discuss in another thread, since they are related

**Comment by derrandz** [comment]
Here's a screenshot of the last changes that shows:

1. In green: the time factor of uptime: `(total node runtime in seconds) / (time() - node start time ts)`
2. In yellow: the work factor of uptime `(total headers sampled) / network head`

Screen Shot 2023-01-31 at 01 10 36

File: nodebuilder/node/metrics.go [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+		)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Observe node start timestamp
+	nodeStartTS.Observe(context.Background(), float64(time.Now().UTC().Unix()))
+
+	// Observe total node run time
+	var totalNodeUpTimeInSeconds int64
```
Comment (walldiss) [inline nodebuilder/node/metrics.go:46]: Looks like this value needs to be init with `time.Now().Unix()` value instead of 0

**Review by walldiss — COMMENTED** [review]
Just one more small issue and it looks good. Could you please add test for DASer sampled amount metric?

File: das/metrics.go [hunk 19]
```diff
@@ -76,16 +80,31 @@ func (d *DASer) InitMetrics() error {
 		return err
 	}
 
+	totalSampled, err := meter.
+		AsyncInt64().
+		Gauge(
+			"das_total_sampled_headers_gauge",
```
Comment (renaynay) [inline das/metrics.go:86]:
```suggestion
			total_sampled_headers",
```

File: das/metrics.go [hunk 20]
```diff
@@ -24,7 +25,10 @@ type metrics struct {
 	sampleTime    syncfloat64.Histogram
 	getHeaderTime syncfloat64.Histogram
 	newHead       syncint64.Counter
-	lastSampledTS int64
+	totalSampled  asyncint64.Gauge
```
Comment (renaynay) [inline das/metrics.go:28]: this looks like it's unused on the struct - why are you storing it? you only need it for the callback func

File: das/metrics.go [hunk 21]
```diff
@@ -110,29 +132,54 @@ func (d *DASer) InitMetrics() error {
...
+// observeNewHead records the network head.
 func (m *metrics) observeNewHead(ctx context.Context) {
 	if m == nil {
 		return
 	}
 	m.newHead.Add(ctx, 1)
 }
+
+// recordTotalSampled records the total sampled headers.
+func (m *metrics) recordTotalSampled(ctx context.Context, totalSampled int) {
```
Comment (renaynay) [inline das/metrics.go:180]: you don't use the context here, do you need to pass it?

File: das/options.go [hunk 22]
```diff
@@ -38,7 +38,8 @@ type Parameters struct {
 	// SampleFrom is the height sampling will start from if no previous checkpoint was saved
 	SampleFrom uint64
 
-	// SampleTimeout is a maximum amount time sampling of single block may take until it will be canceled
+	// SampleTimeout is a maximum amount time sampling of
+	// single block may take until it will be canceled
```
Comment (renaynay) [inline das/options.go:42]: looks weird formatting

File: das/worker.go [hunk 23]
```diff
@@ -60,8 +60,16 @@ func (w *worker) run(
 		}
 
 		metrics.observeGetHeader(ctx, time.Since(startGet))
-		log.Debugw("got header from header store", "height", h.Height(), "hash", h.Hash(),
-			"square width", len(h.DAH.RowsRoots), "data root", h.DAH.Hash(), "finished (s)", time.Since(startGet))
+
+		// TODO(@team): find linter rule to enforce this
```
Comment (renaynay) [inline das/worker.go:64]: there is a linter rule to enforce that a LOC doesn't exceed certain char amount, so you can remove this todo

File: nodebuilder/node/metrics.go [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+	"go.opentelemetry.io/otel/metric/instrument"
+)
+
+var (
+	meter = global.MeterProvider().Meter("node")
+	log   = logging.Logger("node")
+)
+
+// WithMetrics registers node metrics.
+func WithMetrics() {
```
Comment (renaynay) [inline nodebuilder/node/metrics.go:20]: please return error here instead of doing log.Fatal

File: nodebuilder/node/metrics.go [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+			"node_runtime_counter_in_seconds",
+			instrument.WithDescription("total time the node has been running"),
+			instrument.WithUnit("seconds"),
+		)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var started bool = false
+	var totalNodeUpTimeInSeconds int64 // Observe total node run time
```
Comment (renaynay) [inline nodebuilder/node/metrics.go:43]:
```suggestion
	var (
	    started bool
	    totalNodeUptimeInSeconds int64
	 ) 
```

Comment (renaynay) [reply nodebuilder/node/metrics.go:43]: and obviously format pls

File: nodebuilder/node/metrics.go [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+			if !started {
+				// Observe node start timestamp
+				nodeStartTS.Observe(context.Background(), float64(time.Now().UTC().Unix()))
+				started = true
+			}
+
+			now := time.Now().Unix()
+			last := atomic.SwapInt64(&totalNodeUpTimeInSeconds, now)
+
+			totalNodeRunTime.Observe(ctx, float64(now-last))
```
Comment (renaynay) [inline nodebuilder/node/metrics.go:56]: use time.Since

File: nodebuilder/node/metrics.go [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+		},
+	)
+	if err != nil {
+		log.Fatal(err)
```
Comment (renaynay) [inline nodebuilder/node/metrics.go:60]: return err

**Review by renaynay — CHANGES_REQUESTED** [review]

File: das/metrics.go (hunk 21 above)
Comment (derrandz) [reply das/metrics.go:180]: 👍🏻

File: das/worker.go (hunk 23 above)
Comment (derrandz) [reply das/worker.go:64]: I meant more like, the style of having `"attribute", var` per line than the LOC length. Seems that such rule doesn't exist.

File: nodebuilder/node/metrics.go (hunk 26 above)
Comment (derrandz) [reply nodebuilder/node/metrics.go:56]: We agreed me and @walldiss to rely on simple subtraction for simplicity since both values are just "distances" in time in seconds.

File: das/metrics.go (hunk 20 above)
Comment (derrandz) [reply das/metrics.go:28]: will keep ftm to make use of dependency injection during test to inject a mock.

**Comment by derrandz** [comment]
Merge is blocked by #1537

**Comment by mojtaba-esk** [comment]
Hello,
I have to observations to report.

The total runtime shows a floating point value:
`curl -s  http://localhost:9093/api/v1/query?query=node_runtime_counter_in_seconds | jq`
```json
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": [
      {
        "metric": {
          "__name__": "node_runtime_counter_in_seconds",
          "exported_instance": "12D3KooWMuCkRJfEaae4mQ4HPL5FnNe1zc6q8ASCk9MdvEkK5DUj",
          "exported_job": "Celestia-Bridge",
          "instance": "otel-collector:8889",
          "job": "da_nodes"
        },
        "value": [
          1675688581.167,
          "0.00000006"
        ]
      }
    ]
  }
}
```
The `total_synced` metric is not appearing in my Bridge node instance metrics.
`curl -s  http://localhost:9093/api/v1/label/__name__/values | jq`
```json
{
  "status": "success",
  "data": [
    "badencoding",
    "das_busy_workers_amount",
    "das_get_header_time_hist_bucket",
    "das_get_header_time_hist_count",
    "das_get_header_time_hist_sum",
    "das_head_updated_counter",
    "das_latest_sampled_ts",
    "das_network_head",
    "das_sample_time_hist_bucket",
    "das_sample_time_hist_count",
    "das_sample_time_hist_sum",
    "das_sampled_chain_head",
    "das_sampled_headers_counter",
    "das_total_sampled_headers",
    "das_total_sampled_headers_gauge",
    "head",
    "last_pfb_timestamp",
    "node_runtime_counter_in_seconds",
    "node_start_ts",
    "pfb_count",
    "scrape_duration_seconds",
    "scrape_samples_post_metric_relabeling",
    "scrape_samples_scraped",
    "scrape_series_added",
    "target_info",
    "up"
  ]
}
```

File: share/p2p/shrexsub/pubsub.go [hunk 28]
```diff
@@ -39,7 +39,8 @@ type PubSub struct {
 
 // NewPubSub creates a libp2p.PubSub wrapper.
 func NewPubSub(ctx context.Context, h host.Host, suffix string) (*PubSub, error) {
-	// WithSeenMessagesTTL without duration allows to process all incoming messages(even with the same msgId)
+	// WithSeenMessagesTTL without duration allows to
+	// process all incoming messages(even with the same msgId)
```
Comment (renaynay) [inline share/p2p/shrexsub/pubsub.go:43]: why ? pls revert unrelated changes

File: nodebuilder/tests/swamp/swamp_tx.go [hunk 29]
```diff
@@ -2,13 +2,16 @@ package swamp
...
 	"context"
+	"time"
 
 	"github.com/cosmos/cosmos-sdk/client/flags"
 )
 
 // FillBlocks produces the given amount of contiguous blocks with customizable size.
 // The returned channel reports when the process is finished.
 func (s *Swamp) FillBlocks(ctx context.Context, bsize, blocks int) chan error {
+	// TODO @renaynay: figure out why sleep is necessary to prevent flakeyness for macOS
```
Comment (renaynay) [inline nodebuilder/tests/swamp/swamp_tx.go:13]: Please drop commit associated w these changes.

File: .gitignore [hunk 30]
```diff
@@ -20,3 +20,4 @@ vendor
 /cel-key
 coverage.txt
 go.work
+./metals
```
[the remark this answers is not in the document]
Comment (derrandz) [reply .gitignore]: was caused by a weird vscode extension from Scala stuff, removed.

**Comment by derrandz** [comment]
Back to you @mojtaba-esk , @renaynay has been able to provide a few fixes to the issues you've raised.

File: das/checkpoint.go [hunk 31]
```diff
@@ -48,3 +48,12 @@ func (c checkpoint) String() string {
 
 	return str
 }
+
+// totalSampled returns the total amount of sampled headers
+func (c checkpoint) totalSampled() int {
```
Comment (vgonkivs) [inline das/checkpoint.go:53]: What if `totalSampled < 0`?
Can we make the return value `uint64`? It will allow us to avoid converting `uint64` to `int` in this func.  We can convert it into int in`recordTotalSampled`?

File: das/metrics.go [hunk 32]
```diff
@@ -24,7 +25,10 @@ type metrics struct {
 	sampleTime    syncfloat64.Histogram
 	getHeaderTime syncfloat64.Histogram
 	newHead       syncint64.Counter
-	lastSampledTS int64
+	totalSampled  asyncint64.Gauge
+
+	lastSampledTS   int64
+	totalSampledInt int64
```
Comment (vgonkivs) [inline das/metrics.go:31]: Is it necessary to make it `int64`? Or we can also make it `uint64`

File: das/metrics_test.go [hunk 33]
```diff
@@ -0,0 +1,111 @@
...
+	// require.NoError(t, err)
+
+	// totalSampled = mr.ScopeMetrics[0].Metrics[8]
+	// totalSampledGauge, _ = (totalSampled.Data).(metricdata.Gauge[float64])
+
+	// // assert that the collected metrics were the correct onces
+	// assert.Equal(t, totalSampled.Name, "das_total_sampled_headers")
+
+	// // assert for data correctness
+	// assert.Equal(t, len(totalSampledGauge.DataPoints), 2)
```
Comment (vgonkivs) [inline das/metrics_test.go:66]: commented code?

File: das/metrics_test.go [hunk 34]
```diff
@@ -0,0 +1,111 @@
...
+
+	err = daser.Stop(ctx)
+	require.NoError(t, err)
+}
+
+// test recordTotalSampled
+func TestRecordTotalSampled(t *testing.T) {
+	t.Skip()
+}
```
Comment (vgonkivs) [inline das/metrics_test.go:75]: ???

File: libs/header/sync/metrics.go [hunk 35]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"context"
+	"sync/atomic"
+
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+)
+
+var meter = global.MeterProvider().Meter("header/sync")
```
Comment (vgonkivs) [inline libs/header/sync/metrics.go:11]: since we now have two `header` folders, I'd suggest renaming it to `libs/syncer`.
 @renaynay , wdyt?

File: nodebuilder/node/metrics_test.go [hunk 36]
```diff
@@ -0,0 +1,82 @@
...
+	}
+
+	reader := metric.NewPeriodicReader(exp, metric.WithTimeout(1*time.Second))
+
+	provider := metric.NewMeterProvider(
+		metric.WithReader(reader),
+	)
+
+	return provider, reader, nil
+}
```
Comment (vgonkivs) [inline nodebuilder/node/metrics_test.go:82]: We are using the same function in node/metrics_test.go and in das/metrics_test.go. Is it expected to be used somewhere else? If so, can we move it `libs/utils`?

File: libs/header/sync/metrics.go (hunk 35 above)
Comment (renaynay) [reply libs/header/sync/metrics.go:11]: might be hard to differentiate for metrics-sake

File: nodebuilder/node/metrics_test.go (hunk 36 above)
Comment (derrandz) [reply nodebuilder/node/metrics_test.go]: Deprecating for now

File: das/metrics.go (hunk 32 above)
Comment (derrandz) [reply das/metrics.go:29]: I think `uint64` is more fitting

File: das/checkpoint.go (hunk 31 above)
Comment (derrandz) [reply das/checkpoint.go:53]: sounds good

File: das/metrics.go (hunk 32 above)
Comment (derrandz) [reply das/metrics.go:29]: although this will cause conversion when calling `.Observe` since the metric type is `AsyncInt64.Gauge`

File: das/metrics.go (hunk 20 above)
Comment (derrandz) [reply das/metrics.go]: thought I'd need it for testing, turns out not needed, and in either case now tests are removed so...

File: libs/header/sync/metrics.go (hunk 35 above)
Comment (derrandz) [reply libs/header/sync/metrics.go:11]: keeping it as is ftm unless you guys think otherwise

File: das/metrics.go [hunk 37]
```diff
@@ -110,29 +129,54 @@ func (d *DASer) InitMetrics() error {
...
 		return
 	}
 	m.newHead.Add(ctx, 1)
 }
+
+// recordTotalSampled records the total sampled headers.
+func (m *metrics) recordTotalSampled(totalSampled uint64) {
+	if m == nil {
+		return
+	}
```
Comment (renaynay) [inline das/metrics.go:180]:
```suggestion
```

please instead ensure that syncer never calls any method on a nil metrics.

File: das/metrics.go [hunk 38]
```diff
@@ -24,7 +24,9 @@ type metrics struct {
 	sampleTime    syncfloat64.Histogram
 	getHeaderTime syncfloat64.Histogram
 	newHead       syncint64.Counter
-	lastSampledTS int64
+
+	lastSampledTS   int64
```
Comment (renaynay) [inline das/metrics.go:28]: fwiw this could also be uint64 for consistency

File: libs/header/sync/sync.go [hunk 39]
```diff
@@ -212,7 +214,9 @@ func (s *Syncer[H]) sync(ctx context.Context) {
 	log.Infow("finished syncing",
 		"from", s.syncedHead.Height(),
 		"to", newHead.Height(),
-		"elapsed time", s.state.End.Sub(s.state.Start))
+		"elapsed time",
+		s.state.End.Sub(s.state.Start),
```
Comment (renaynay) [inline libs/header/sync/sync.go:218]:
```suggestion
		"elapsed time", s.state.End.Sub(s.state.Start),
```

clearer

File: nodebuilder/tests/swamp/swamp_tx.go [hunk 40]
```diff
@@ -9,6 +9,8 @@ import (
 // FillBlocks produces the given amount of contiguous blocks with customizable size.
 // The returned channel reports when the process is finished.
 func (s *Swamp) FillBlocks(ctx context.Context, bsize, blocks int) chan error {
+	// wait for first block to begin filling blocks
+	s.WaitTillHeight(ctx, 1)
```
Comment (renaynay) [inline nodebuilder/tests/swamp/swamp_tx.go:13]: please remove these lines bc i'll add it in eventually with a diff pr

**Review by renaynay — CHANGES_REQUESTED** [review]

File: das/metrics.go (hunk 37 above)
Comment (derrandz) [reply das/metrics.go:180]: Do you mean `DASer`? We had a conversation me and @Wondertan about this pattern, and I mentioned that we should remove it.

For me to change this `if m == nil` would mean a major refactor to how we are doing metrics atm for DASer. The "better" pattern that we should have, that makes sense that metrics collection methods are not called in the first place when `Metrics` are disabled, is present in the blackbox metrics PR #1376.

To change this, we need a non-trivial refactor

Comment (derrandz) [reply das/metrics.go:180]: Referencing this in https://github.com/celestiaorg/celestia-node/issues/1705

File: das/metrics.go (hunk 38 above)
Comment (derrandz) [reply das/metrics.go:28]: noted, (_although not related to this PR_)

File: das/metrics.go (hunk 37 above)
Comment (renaynay) [reply das/metrics.go:180]: Since you already implemented the check here: https://github.com/celestiaorg/celestia-node/pull/1638/files#diff-a79f038a558af1c9917378c69183fb2563c3a0b6f49ae9464b1bfae68d2f1339R238-R240

we no longer need this check anyway

**Review by renaynay — DISMISSED** [review]

File: nodebuilder/node_test.go [hunk 41]
```diff
@@ -47,3 +55,83 @@ func TestLifecycle(t *testing.T) {
...
+		rawResponse, _ := proto.Marshal(&response)
+		contentType := "application/x-protobuf"
+		status := http.StatusOK
+
+		log.Debug("Responding to optl POST request")
+		w.Header().Set("Content-Type", contentType)
+		w.WriteHeader(status)
+		_, _ = w.Write(rawResponse)
+
+		log.Debug("Responded to optl POST request")
```
Comment (distractedm1nd) [inline nodebuilder/node_test.go:132]:
```suggestion
		log.Debug("Responding to otlp POST request")
		w.Header().Set("Content-Type", contentType)
		w.WriteHeader(status)
		_, _ = w.Write(rawResponse)

		log.Debug("Responded to otlp POST request")
```

**Review by distractedm1nd — APPROVED** [review]

**Review by distractedm1nd — APPROVED** [review]

File: das/checkpoint.go [hunk 42]
```diff
@@ -48,3 +48,12 @@ func (c checkpoint) String() string {
...
 	return str
 }
+
+// totalSampled returns the total amount of sampled headers
+func (c checkpoint) totalSampled() uint64 {
+	totalInProgress := 0
+	for _, w := range c.Workers {
+		totalInProgress += int(w.To-w.From) + 1
+	}
+	return uint64(int(c.SampleFrom) - totalInProgress - len(c.Failed))
```
Comment (vgonkivs) [inline das/checkpoint.go:58]:
```suggestion
	var totalInProgress uint64
	for _, w := range c.Workers {
		totalInProgress += w.To-w.From + 1
	}
	return c.SampleFrom - totalInProgress - uint64(len(c.Failed)))
```

**Review by vgonkivs — COMMENTED** [review]
Small nits regarding type conversion but other than that lgtm

**Review by distractedm1nd — APPROVED** [review]

**Review by renaynay — APPROVED** [review]

**Review by vgonkivs — APPROVED** [review]
